packages/ckeditor5-bookmark/src/insertbookmarkcommand.ts

```diff
@@ -0,0 +1,127 @@
+/**
+ * @license Copyright (c) 2003-2024, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module bookmark/insertbookmarkcommand
+ */
+
+import type { DocumentSelection, Item, Element, Selection } from 'ckeditor5/src/engine.js';
+import { Command } from 'ckeditor5/src/core.js';
+
+/**
+ * The insert bookmark command.
+ *
+ * The command is registered by {@link module:bookmark/bookmarkediting~BookmarkEditing} as `'insertBookmark'`.
+ *
+ * To insert a bookmark element at place where is the current collapsed selection or where is the beginning of document selection,
+ * execute the command passing the bookmark id as a parameter:
+ *
+ * ```ts
+ * editor.execute( 'insertBookmark', { bookmarkId: 'foo_bar' } );
+ * ```
+ */
+export default class InsertBookmarkCommand extends Command {
+	/**
+	 * @inheritDoc
+	 */
+	public override refresh(): void {
+		const model = this.editor.model;
+		const selection = model.document.selection;
+		const startPosition = selection.isCollapsed ? selection.anchor! : selection.getFirstPosition()!;
+
+		const isBookmarkAllowedBySchema = model.schema.checkChild( startPosition, 'bookmark' );
+		const isParagraphAllowedBySchema = model.schema.checkChild( startPosition, 'paragraph' );
+		const isAnyAllowedElementInSelection = !!this._getAllowedElementBasedOnSelection( selection );
+
+		this.isEnabled = isBookmarkAllowedBySchema || isParagraphAllowedBySchema || isAnyAllowedElementInSelection;
+	}
+
+	/**
+	 * Executes the command.
+	 *
+	 * @fires execute
+	 * @param options Command options.
+	 * @param options.bookmarkId The value of the `bookmarkId` attribute.
+	 */
+	public override execute( options: { bookmarkId: string } ): void {
+		if ( !options ) {
+			return;
+		}
+
+		const { bookmarkId } = options;
+
+		if ( !bookmarkId || typeof bookmarkId !== 'string' ) {
+			return;
+		}
+
+		const editor = this.editor;
+		const model = editor.model;
+		const selection = model.document.selection;
+
+		model.change( writer => {
+			const startPosition = selection.isCollapsed ? selection.anchor! : selection.getFirstPosition()!;
+			const isBookmarkAllowedBySchema = model.schema.checkChild( startPosition, 'bookmark' );
+
+			// If it is allowed to insert bookmark by schema than insert it.
+			if ( isBookmarkAllowedBySchema ) {
+				writer.setSelection( startPosition );
+
+				return model.insertObject( writer.createElement( 'bookmark', { bookmarkId } ) );
+			} else {
```

Comment:
I'd suggest refactoring. Maybe a helper can help find the insertion position. If the position can not be found then the command should be disabled (common logic for execute and refresh). If the returned position does not allow for bookmark but allows for a paragraph then insert a paragraph with a bookmark inside.